example_updates/rollback_update/example-rollback-update.sh

```diff
@@ -0,0 +1,798 @@
+                #
+                # Here is how we generated the private key for signing the image
+                # and how we generated that public key file used to validate the image signature.
+                #
+                # Generated RSA private key with password using command:
+                # openssl genrsa -aes256 -passout file:priv.pass -out priv.pem
+                #
+                # Generated RSA public key from private key using command:
+                # openssl rsa -in ${WORKDIR}/priv.pem -out ${WORKDIR}/public.pem -outform PEM -pubout
+
+                ret_val=1
+                if [[ -f ${public_key_file} ]]; then
+                    # Call swupdate with the image file and the public key for signature validation
+                    swupdate -v -i "${image_file}" -k "${public_key_file}" -e ${selection} &>> "${swupdate_log_file}"
+                    ret_val=$?
+                else
+                    # Call swupdate with the image file and no signature validations
+                    swupdate -v -i "${image_file}" -e ${selection} &>> "${swupdate_log_file}"
+                    ret_val=$?
+                fi
+
+                if [[ $ret_val -eq 0 ]]; then
+                    resultCode=600
+                    extendedResultCode=0
+                    resultDetails=""
+                else
+                    resultCode=0
+                    make_swupdate_handler_erc "$ret_val" extendedResultCode
+                    resultDetails="SWUpdate command failed."
+                fi
+            else
+                echo "Image file ${image_file} was not found." >> "${log_file}"
+                resultCode=0
+                # ADUC_ERC_SWUPDATE_HANDLER_INSTALL_FAILURE_IMAGE_FILE_NOT_FOUND (0x202)
+                extendedResultCode=0x30100202
+                resultDetails="Image file ${image_file} was not found."
+            fi
+            ret_val=0
+        else
+            # Installed.
+            log_info "It appears that this component already installed the specified update."
+            resultCode=603
+            extendedResultCode=0
+            resultDetails="Already installed."
+            ret_val=0
+        fi
+    fi
+   
+
+    # Prepare ADUC_Result json.
+    make_aduc_result_json "$resultCode" "$extendedResultCode" "$resultDetails" aduc_result
+
+    # Show output.
+    output "Result:" "$aduc_result"
+
+    # Write ADUC_Result to result file.
+    result "$aduc_result"
+
+    $ret $ret_val
+}
+
+#
+# Performs steps to finalize the update.
+#
+ApplyUpdate() {
+    log_info "Applying update..."
+
+    ret_val=0
+    resultCode=0
+    extendedResultCode=0
+    resultDetails=""
+
+    curr_workflow=$(basename $workfolder)
+
+    echo "Applying." >> "${log_file}"
+
+    # Set the bootloader environment variable
+    # to tell the bootloader to boot into the new partition.
+    #
+    # 'rpipart' variable is specific to our boot.scr script.
+    
+    fw_setenv rpipart $update_partition
```

Comment:
is semantics of rpipart still "commit before boot attempt" or do we commit after successfully booting and passing IsInstalled() -> true check on first boot into new image?  what u-boot variable(s) are used to determine what the current committed partition is to use for steady state?

Reply:
Steady state will be n-1 rpipart. So for rpipart 3 we would fall back to rpipart 4 and vice versa. 

We are committing to move to the next partition hear. We handle the rollback within the actual u-boot script here: https://github.com/Azure/meta-raspberrypi-adu/blob/user/nihemstr/rollback-impl/recipes-bsp/rpi-u-boot-scr/files/boot.cmd.in